### ISSUE TITLE
final TypedDict

### ISSUE THREAD

**Issue by crusaderky** [python/mypy#7981]
Reopened from #7845 as requested

mypy seems to ignore the ``@final`` decorator when it is applied to a TypedDict. This issue becomes material when one invokes the items() or values() methods, or passes a generic string to get(); then the output will be incorrectly upcast to object since mypy thinks it may be dealing with a subclass.

```python
from typing_extensions import final, TypedDict

@final
class Point(TypedDict):
    x: float
    y: float
    z: float


p: Point
x: str
reveal_type(p.values())
reveal_type(p.items())
reveal_type(p.get(x, 0.0))

```
Output:
```
12: note: Revealed type is 'typing.ValuesView[builtins.object*]'
13: note: Revealed type is 'typing.AbstractSet[Tuple[builtins.str*, builtins.object*]]'
14: note: Revealed type is 'builtins.object*'
```
Expected output:
```
12: note: Revealed type is 'typing.ValuesView[builtins.float*]'
13: note: Revealed type is 'typing.ItemsView[builtins.str*, builtins.float*]'
14: note: Revealed type is 'builtins.float*'
```

**Comment by crusaderky** [comment]
Related: #7849, #7865

**Comment by JukkaL** [comment]
The proposed semantics seem to be that a final TypedDict object doesn't have any extra keys beyond those included in the definition. Also, a final TypedDict can't be used as a base to define derived TypedDicts.

I haven't thought about this carefully, but it may be possible to define this in a sound fashion. A final TypedDict would only be compatible with another final TypedDict, and they must have the same keys and required keys, and the key types must be compatible.

Since this seems to work at runtime, this could be implemented without changes to `typing`. However, I'm not sure how useful this would be. I'd be interested in hearing if anybody has real-world use cases where this would be helpful.

**Comment by ilevkivskyi** [comment]
> A final TypedDict would only be compatible with another final TypedDict, and they must have the same keys and required keys, and the key types must be compatible.

Exactly, it looks like this is the only way to make this sound. This however may be not very useful, as one would need to have same typed dict all the way down the call stack. On the other hand this may be OK for _new_ codebases.

**Comment by crusaderky** [comment]
The same problem is on the keys. ``.keys()`` and ``.__iter__()`` yield ``str`` according to mypy, but the final decorator could make it change that to ``Literal``.

Real life use case:

```python
from typing_extensions import TypedDict, final


@final
class Counters(TypedDict):
    counter_1: int
    blah_blah: int
    something_else: int
    and_another: int
    one_more: int


def reset_counters(c: Counters) -> None:
    for k in c:
        c[k] = 0  # error: TypedDict key must be a string literal; expected one of ('counter_1', 'blah_blah', 'something_else', 'and_another', 'one_more')
```

**Comment by intgr** [comment]
I would find the implementation propsed by @ilevkivskyi quite useful. I have a TypedDict whose all values are `List[something]` and I would like to iterate over the dict's `items()` in a generic manner.

**Comment by JosiahKane** [comment]
I would also find this useful in annotating real world code, as indicated in the linked closed issue. 

My actual usecase is a Computer Vision type problem where we're using TypedDict to package together corresponding images. For example, there might be an RGB image and a binary mask indicating the foreground of that image.  

```
class Image:
    def save(self, filename: str) -> None:
        ...

class ColourImage(Image):
    ...

class BitmaskImage(Image):
    ...

@final
class ImageAnnotation(TypedDict):
    img: ColourImage
    foreground: BitmaskImage
    # Although these are all Images, we don't want to use a general Dict[str, Image] because we'd lose the specialisations.

def save_all(img_pack: ImageAnnotation, prefix) -> None:
    for k, v in img_pack.items():
        # Currently this fails because save is not defined for object. 
        v.save(f"{prefix}_{k}.png")
        # It would instead be necessary to say 
        # for k, v in cast(Mapping[str, Image], img_pack).items():

```

A particularly elegant construction that might be enabled by marking a `TypedDict` closed for extension is the ability to create a new one of the same kind with a comprehension. For example

```
ImageT = TypeVar("ImageT", bound=Image)

def resize(img: ImageT, scale: float) -> ImageT:
    ...

def resize_all(img_pack: ImageAnnotation, scale: float) -> ImageAnnotation:
    return {k: resize(v, scale) for k, v in img_pack.items()}
    # This wouldn't work even with a cast
```

**Comment by wmdrichards** [comment]
I've also just run into this issue -- use case is I have a number of TypedDicts that are all Mapping[str, T], but I'd like to restrict the keys to a known set. Besides declaring `@final`, one other option that comes to mind is to make TypedDict a generic class, so that the following syntax would work:
```
class Point(TypedDict[float]):
    x: float
    y: float
    z: float
```
This would _I think_ solve the use cases in this thread, without disallowing extension

**Comment by smurzin** [comment]
This now also makes mypy diverge with pylance - https://github.com/microsoft/pyright/issues/1899
They've decided to allow final for `TypedDict` in order to be able to provide type narrowing in `Union[TypedDict]` case.

Which is problematic for projects that use pylance for IDE/intellisense and mypy in CI.

**Comment by 97littleleaf11** [comment]
Jukka‘s comment about this: https://github.com/python/mypy/issues/12266#issue-1155535982

**Comment by godlygeek** [comment]
> pylance ... decided to allow final for `TypedDict` in order to be able to provide type narrowing in `Union[TypedDict]` case.

I find myself in exactly this case. Imagine I'm hitting an API that returns a JSON response. If my request succeeded, the service responds with something like
```py
{"transactionDetails": {"transactionId": "1234-56-7890"}}
```
but if my request failed the service response with something like
```py
{"error": {"code": 1234, "description": "Oops!"}}
```
I can annotate each of those possibilities as a `TypedDict`:
```py
class TransactionDetails(TypedDict):
    transactionId: str

class TransactionDetailsResponse(TypedDict):
    transactionDetails: TransactionDetails

class Error(TypedDict):
    code: int
    description: str

class ErrorResponse(TypedDict):
    error: Error

Response = ErrorResponse | TransactionDetailsResponse
```

At this point, what I'd like mypy to understand is that something typed as `Response` either has a `"transactionDetails"` key (in which case its type should narrow to `TransactionDetailsResponse`) or it has a `"error"` key (in which case its type should narrow to `ErrorResponse`).

Unfortunately, I'm instead stuck with casts:
```py
def handle_response(response: Response):
    if "transactionDetails" in response:
        print(f"Transaction {cast(TransactionDetailsResponse, response)['transactionDetails']['transactionId']} submitted!")
    else:
        print(f"Got error {cast(ErrorResponse, response)['error']['code']}: {cast(ErrorResponse, response)['error']['description']}")
```

I really don't want that cast. I want to say that `"transactionDetails" in response` is a `TypeGuard[TransactionDetailsResponse]`, and that seems to be exactly what pyright implemented. It's not onerous to me to mark these `TypedDict`'s as `Final`, since they _are_ required to have distinct keys in order for me to distinguish which type of response I got - that's a part of the service's API contract, and it doesn't make sense for there to be subtypes for exactly the same reason.

**Comment by Hnasar** [comment]
> the output will be incorrectly upcast to object since mypy thinks it may be dealing with a subclass.

I ran into this limitation and realized mypy's behavior (with assuming `object) is consistent with _PEP 589 - TypedDict_ and the [TypedDict Typing Spec](https://typing.readthedocs.io/en/latest/spec/typeddict.html#typeddict).

> A TypedDict with all int values is not consistent with Mapping[str, int], since there may be additional non-int values not visible through the type, due to structural subtyping. These can be accessed using the values() and items() methods in Mapping, for example.

Currently the above Typing Spec page, and the one for [@final](https://typing.readthedocs.io/en/latest/spec/qualifiers.html#final) don't clarify the interaction between these two features.

**Comment by erictraut** [comment]
> Currently the above Typing Spec page, and the one for [@final](https://typing.readthedocs.io/en/latest/spec/qualifiers.html#final) don't clarify the interaction between these two features.

The `@final` class decorator indicates that a class cannot be subclassed. This makes sense for classes that define nominal types. However, `TypedDict` is a structural type, similar to a `Protocol`. That means two `TypedDict` classes with different names but the same field definitions are equivalent types. Their names and hierarchies don't matter for determining type consistency. For that reason, `@final` has no impact on a `TypedDict` type consistency rules, nor should it change the behavior of `items` or `values`.

What you're looking for is a new concept referred to as a "closed" `TypedDict`. This concept is introduced in draft [PEP 728](https://peps.python.org/pep-0728/). It allows one to specify that a `TypedDict` cannot have any extra fields beyond the ones that are defined. (Alternatively, it allows for additional extra fields that are constrained to a particular type.)

**Comment by alicederyn** [comment]
Should this be closed now PEP 728 has been accepted? I think it's superseded by #18176, assuming we're agreed that `@final` shouldn't be a synonym for `closed=True`.

**Comment by ilevkivskyi** [comment]
Let's keep this open until https://github.com/python/mypy/pull/21382 (or its equivalent split in parts) is merged, just in case.

**Comment by ilevkivskyi** [comment]
Just to confirm for more visibility: `@final` decorator on TypedDicts is now soft-deprecated (and is only supported on the best effort basis), for all relevant use cases, please use PEP 728 `closed=True`.
